walk.go

```diff
@@ -36,65 +36,68 @@ type walkObserver func(info walkObserverInfo) error
 
 // walk determine if in is a `map[string]interface{}` or a `Slice` and traverse it if so, otherwise will
 // treat it as a scalar and invoke the walk observer on the input value directly.
-func walk(in interface{}, expandPaths bool, wo walkObserver) error {
-	switch in.(type) {
-	case map[string]interface{}:
-		return walkMap(in.(map[string]interface{}), expandPaths, wo)
-	case []interface{}:
-		return walkSlice(in.([]interface{}), expandPaths, wo)
+func walk(inVal reflect.Value , expandPaths bool, wo walkObserver) error {
+	switch inVal.Kind()  {
+	case reflect.Map:
+		return walkMap(inVal, expandPaths, wo)
+	case reflect.Slice:
+		return walkSlice(inVal, expandPaths, wo)
 	default:
-		return walkInterface(in, expandPaths, wo)
+		return walkInterface(inVal, expandPaths, wo)
 	}
 }
 
 // walkmap[string]interface{} is a shorthand way to walk a tree with a map as the root.
-func walkMap(m map[string]interface{}, expandPaths bool, wo walkObserver) error {
-	return walkFullMap(m, m, llpath.Path{}, expandPaths, wo)
+func walkMap(mVal reflect.Value, expandPaths bool, wo walkObserver) error {
+	return walkFullMap(mVal, mVal, llpath.Path{}, expandPaths, wo)
 }
 
 // walkSlice walks the provided root slice.
-func walkSlice(s []interface{}, expandPaths bool, wo walkObserver) error {
-	return walkFullSlice(s, map[string]interface{}{}, llpath.Path{}, expandPaths, wo)
+func walkSlice(sVal reflect.Value, expandPaths bool, wo walkObserver) error {
+	return walkFullSlice(sVal, reflect.ValueOf(map[string]interface{}{}), llpath.Path{}, expandPaths, wo)
 }
 
-func walkInterface(s interface{}, expandPaths bool, wo walkObserver) error {
+func walkInterface(s reflect.Value, expandPaths bool, wo walkObserver) error {
 	return wo(walkObserverInfo{
 		value: s,
 		key:   llpath.PathComponent{},
-		root:  map[string]interface{}{},
+		rootVal:  reflect.ValueOf(map[string]interface{}{}),
 		path:  llpath.Path{},
 	})
 }
 
-func walkFull(o interface{}, root map[string]interface{}, path llpath.Path, expandPaths bool, wo walkObserver) (err error) {
+func walkFull(oVal, rootVal reflect.Value, path llpath.Path, expandPaths bool, wo walkObserver) (err error) {
+
+	// Unpack any wrapped interfaces
+	for oVal.Kind() == reflect.Interface {
+		oVal = reflect.ValueOf(oVal.Interface())
+	}
```

Comment:
what about pointers? What if you have nested pointers and interfaces?

some helpers from go-ucfg:
```
func chaseValueInterfaces(v reflect.Value) reflect.Value {
	for v.Kind() == reflect.Interface && !v.IsNil() {
		v = v.Elem()
	}
	return v
}

func chaseValuePointers(v reflect.Value) reflect.Value {
	for v.Kind() == reflect.Ptr && !v.IsNil() {
		v = v.Elem()
	}
	return v
}

func chaseValue(v reflect.Value) reflect.Value {
	for (v.Kind() == reflect.Ptr || v.Kind() == reflect.Interface) && !v.IsNil() {
		v = v.Elem()
	}
	return v
}

func chaseTypePointers(t reflect.Type) reflect.Type {
	for t.Kind() == reflect.Ptr {
		t = t.Elem()
	}
	return t
}
```

Reply:
That's a good point (no pun intended).

I think there's an argument that the pointer reflects the actual value. You may actually want to check pointer equality in other words. That said, it really gets thorny here. I feel like interfaces are the right line for this library given that our main goal here is getting to something JSON-ish. 

That said, I may not be seeing a lot of use cases. Does the above line of thinking change still leave you thinking that we should chase pointers? I definitely see how some people might want them.

Reply:
Thinking about this more, I think you may be right about chasing pointers. I think a goal of this lib should be to approximate JSON-like structures, and in JSON all types are nullable, which means pointers should be able to go anywhere. 

Let me think about this for a bit, but I suspect you're right.

Reply:
Having thought about it, I think you're right, I added in pointer chasing using `ChaseValue`.